contracts/governance/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+    }
+
+    pub fn new_proposal(
+        env: Env,
+        participant: Address,
+        id: u64,
+        kind: ProposalType,
+        new_contract_hash: BytesN<32>,
+    ) -> Result<(), Error> {
+        participant.require_auth();
+
+        let storage = env.storage().persistent();
+        let mut participant_repo = participant::Repository::new(&storage)?;
+
+        let stored_participant = participant_repo.find(participant.clone())?;
+
+        if !stored_participant.is_whitelisted() {
+            return Err(Error::ParticipantNotWhitelisted);
+        }
+
+        let voting_address = storage
+            .get::<_, Address>(&DataKey::VotingContractAddress)
+            .unwrap();
+
+        let voting_client = voting_contract::Client::new(&env, &voting_address);
+
+        voting_client.create_proposal(&participant, &id, &kind, &new_contract_hash);
+
+        Ok(())
+    }
+
+    pub fn vote(env: Env, participant: Address, id: u64) -> Result<(), Error> {
+        participant.require_auth();
+
+        let storage = env.storage().persistent();
+        let mut participant_repo = participant::Repository::new(&storage)?;
+
+        let stored_participant = participant_repo.find(participant.clone())?;
+
+        if !stored_participant.is_whitelisted() {
+            return Err(Error::ParticipantNotWhitelisted);
+        }
+
+        let voting_address = storage
+            .get::<_, Address>(&DataKey::VotingContractAddress)
+            .unwrap();
+
+        let voting_client = voting_contract::Client::new(&env, &voting_address);
+
+        voting_client.vote(&participant, &id);
+
+        Ok(())
+    }
+
+    pub fn execute_proposal(env: Env, participant: Address, id: u64) -> Result<(), Error> {
+        participant.require_auth();
+
+        let storage = env.storage().persistent();
+        let mut participant_repo = Repository::new(&storage)?;
+
+        let stored_participant = participant_repo.find(participant.clone())?;
+
+        if !stored_participant.is_whitelisted() {
+            return Err(Error::ParticipantNotWhitelisted);
+        }
+
+        let voting_address = storage
+            .get::<_, Address>(&DataKey::VotingContractAddress)
+            .unwrap();
+
+        let voting_client = voting_contract::Client::new(&env, &voting_address);
+
+        let whitelisted_balance = participant_repo.whitelisted_balance(&env);
+
+        let proposal = voting_client.find_proposal(&id);
+
+        if proposal.proposer != participant {
+            return Err(Error::OnlyAuthorCanExecuteProposals);
+        }
+
+        if Self::is_proposal_executed(&storage, id) {
+            return Err(Error::AlreadyExecuted);
+        }
+
+        if !voting_client.is_proposal_approved_for_balance(&proposal.id, &whitelisted_balance) {
+            return Err(Error::ProposalNeedsApproval);
+        }
+
+        match proposal.kind {
+            ProposalType::Standard => {
+                // TODO - should we do anything for standard proposal ?
```

Comment:
Should we do something here ? @geofmureithi @mariopil

Reply:
What is actually `standard` type? When could it be used? Maybe post an additional event that the standard proposal was executed? But on the other hand that could overlap with the event posted at the end of execution function...

Reply:
> What is actually standard type?

Thats a good question. I used that type as a placeholder to say "well, here we maybe should be doing something more". Maybe `Standard` is not a good name and we could just use "Votation", trying to signal we are just voting a proposal, but that it has not further effects or invocations.